### PR TITLE
cuda: use PCIe P2P copies when migrating between cuda devices

### LINKED ISSUE #1187: cuda: consider making use of cuMemcpyPeerAsync for D2D migrations
Requires checking if the PCIe topology allows P2P copies with `cudaDeviceCanAccessPeer` and maybe caching a list of peers in each device so the check does not need to be repeated with each migration.

### PR DESCRIPTION
**Pull request by jansol**
In which the [improved FluidX3D](https://github.com/ProjectPhysX/FluidX3D/tree/experimental-p2p) runs on an 8 GPU setup* 3.6x as fast as with the Nvidia proprietary OpenCL driver.

It seems Nvidia's driver takes a huge performance hit if buffers are migrated between devices, as the previous FluidX3D release (2.4) runs much better despite doing buffer migrations manually by downloading buffer contents to host memory and reuploading to the target device. Still, even compared to the old approach this new FluidX3D running on PoCL-CUDA with this patch applied is 2.0x faster.

In other words, this massively improves multi-GPU scaling of PoCL-CUDA (but does not affect single GPU performance in any way).

Some unrelated fixes to PoCL-CUDA are still needed to make FluidX3D interactive graphics work but benchmark mode and rendering to disk works.

Closes #1187

*8x RTX A6000, FluidX3D in benchmark mode the memory limit set to 8192MB.

### CODE REVIEW

**Comment by pjaaskel** [comment]
Nice! Talking about performance low hanging fruits! Did you run clang-format? Also, does it update PoCL event time stamps?

**Comment by pjaaskel** [comment]
Also please update CHANGES.

**Comment by jansol** [comment]
Formatted & added note

**Comment by jansol** [comment]
The code path is the same as for all other commands (minus the d2d migration function itself), so timestamps get updated by the shared code that calls into the D2D migration function and handles command completion later.

File: lib/CL/devices/cuda/pocl-cuda.c [hunk 1]
```diff
@@ -802,6 +802,34 @@ pocl_cuda_free (cl_device_id device, cl_mem mem_obj)
   p->version = 0;
 }
 
+int
+pocl_cuda_can_migrate_d2d (cl_device_id dest, cl_device_id source)
+{
+  assert (dest != source);
+  if (strcmp (source->ops->device_name, source->ops->device_name) == 0)
```
Comment (isuruf) [inline lib/CL/devices/cuda/pocl-cuda.c:809]: Just for my understanding, does this mean you can't migrate from two CUDA devices with different names?

Comment (jansol) [reply lib/CL/devices/cuda/pocl-cuda.c]: And while replying to this I noticed I had copied & pasted "source" there twice. :laughing:  Fixed.
